app/models/user.rb

```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save {self.email = email.downcase}
+  validates :name, presence: true, length: {maximum: Settings.max_name_length}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: Settings.max_email_length},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: {case_sensitive: false}
+  has_secure_password
```

Comment:
dòng này để làm gì thế e?

Reply:
`has_secure_password` cái này đại loại nó sẽ tạo ra 2 thuộc tính `password` với `password_confirmation`. Nó sẽ đảm bảo password và confirm khớp nhau sau đó băm password cùng với salt bằng thuật toán `bcrypt`, sau cùng nó lưu đoạn chuỗi đã băm đó vào `password_digest` trong CSDL ạ.

Reply:
À còn tác dụng là để xác thực password của người dùng chính xác, trong khi không phải lưu password dưới dạng plain text vào CSDL (nguy cơ khi CSDL bị hack sẽ làm lộ password của người dùng) ạ.

Reply:
có thuật toán nào tên là `bcrypt` hả em?

Reply:
@DuongNT1512 Dạ đúng hơn thì có gọi là hàm băm mật khẩu ạ.